### PR TITLE
Prevent oob access in asyncFile.readLine

### LINKED ISSUE #10641: Error when reading asynchronous text file line by line
When I tried to read a text file asynchronous line by line it gives an error and only reads the first line.

### Example

```nim
import asyncdispatch, asyncfile, os

proc main() {.async.} =
  var filename = "test.txt"
  var file = openAsync(filename, fmRead)
  while true:
    let line = await file.readLine()
    if line.len == 0: break
    echo line
  
  file.close()

waitFor main()
```

### Current Output
```
Exception message: index out of bounds
Exception type: [IndexError]
Error: execution of an external program failed: '/home/hdias/Downloads/tmp/test '
```

### Additional Information

```
$ nim -v
Nim Compiler Version 0.19.4 [Linux: amd64]
Compiled at 2019-02-01
Copyright (c) 2006-2018 by Andreas Rumpf

git hash: b6d96cafc8bcad1f3d32f2910b25cd11a93f7751
active boot switches: -d:release
```

### PR DESCRIPTION
**Pull request by LemonBoy**
Fixes #10641

### CODE REVIEW

**Comment by dom96** [comment]
This needs to match the behaviour of ``File``. IIRC ``File.readLine`` will raise an exception.

**Comment by LemonBoy** [comment]
Done. `EOFError` is now raised when one tries to read past the end.

File: lib/pure/asyncfile.nim [hunk 1]
```diff
@@ -300,14 +300,18 @@ proc readLine*(f: AsyncFile): Future[string] {.async.} =
...
   result = ""
   while true:
-    var c = await read(f, 1)
+    let c = await read(f, 1)
+    if c.len == 0:
+      # EOF
+      raise newException(EOFError, "EOF reached")
     if c[0] == '\c':
-      c = await read(f, 1)
+      # Found a CR, eat the NL
```
Comment (dom96) [inline lib/pure/asyncfile.nim]: \L isn't a given here. You can't just discard it and assume that is the case.

Comment (LemonBoy) [reply lib/pure/asyncfile.nim]: And? The PR is meant to fix the `IndexError`, I don't care enough about async files to fix this shit

**Comment by mrhdias** [comment]
I don't understand one thing! Why readLine raise an exception and readAll not?

readAll
```
proc readAll*(f: AsyncFile): Future[string] {.async.} =
  ## Reads all data from the specified file.
  result = ""
  while true:
    let data = await read(f, 4000)
    if data.len == 0:
      return
    result.add data
```

readLine
```
proc readLine*(f: AsyncFile): Future[string] {.async.} =
  ## Reads a single line from the specified file asynchronously.
  result = ""
  while true:
    let c = await read(f, 1)
    if c.len == 0:
      # EOF
      raise newException(EOFError, "EOF reached")
    if c[0] == '\c':
      # Found a CR, eat the NL
      discard await read(f, 1)
      break
    if c[0] == '\L':
      # Found a NL
      break
    result.add(c)
```

File: lib/pure/asyncfile.nim (hunk 1 above)
Comment (dom96) [reply lib/pure/asyncfile.nim]: Fixing one bug and introducing another isn't productive.

Comment (mrhdias) [reply lib/pure/asyncfile.nim]: > Fixing one bug and introducing another isn't productive.

What is your implementation" for the readLine function?

**Comment by mrhdias** [comment]
Why not simplify and do the same thing that Python does?
The user then does what he understands with the new line.

Example

File: test.txt
```
test1
test2

test3
```

Example in python:
```
f = open("test.txt")
for line in f:
    print(line)
f.close()
```
Python output:
```
test1

test2



test3

```
My sugestion for readLine in Nim:
```
import asyncdispatch, asyncfile, os

proc readLine*(f: AsyncFile): Future[string] {.async.} =
    result = ""
    while (let c = await read(f, 1); c.len) > 0:
        result.add(c)
        if c[0] == '\L':
            break


proc main() {.async.} =
    var fh = openAsync("test.txt", fmRead)
    while (let line = await fh.readLine(); line.len) > 0:
        echo line
    fh.close()

waitFor main()
```
Nim output:
```
test1

test2



test3

```

File: lib/pure/asyncfile.nim (hunk 1 above)
Comment (LemonBoy) [reply lib/pure/asyncfile.nim]: > Fixing one bug and introducing another isn't productive.

Oh come the fuck on, you're not even trying.

```nim
  while true:
    var c = await read(f, 1)
    if c[0] == '\c':
      c = await read(f, 1)
      break
    if c[0] == '\L' or c == "":
      break
    else:
      result.add(c)
```

...

```nim
    if c[0] == '\c':
      c = await read(f, 1)
      break
```

...

```nim
break
```

> **introducing another**

Man, you're the one who added this, step up your blame game :point_right: :sunglasses: :point_right:

Comment (dom96) [reply lib/pure/asyncfile.nim]: Fair enough. I misread the code, my bad. But there is absolutely no need to start cursing.

**Comment by dom96** [comment]
@mrhdias because this would be a breaking change. Every ``readLine`` in Nim doesn't include the newline. We could definitely make a case to change this though, but yeah, would be pretty painful.

**Pull request closed without merging**